### PR TITLE
Expose the lock retry count and timout so processes can change it

### PR DESCRIPTION
**Pull request by crmann1**

### CODE REVIEW

File: src/Microsoft.Identity.Client.Extensions.Adal/AdalCache.cs [hunk 1]
```diff
@@ -148,8 +163,10 @@ internal void AfterAccessNotification(TokenCacheNotificationArgs args)
             finally
             {
                 _logger.TraceEvent(TraceEventType.Information, /*id*/ 0, $"Releasing lock");
-                _cacheLock?.Dispose();
+                var localLockCopy = _cacheLock;
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Client.Extensions.Adal/AdalCache.cs]: Please add a comment about the race condition, as it is not straight forward.

File: src/Shared/StorageCreationPropertiesBuilder.cs [hunk 2]
```diff
@@ -72,6 +77,29 @@ public StorageCreationPropertiesBuilder WithMacKeyChain(string serviceName, stri
             return this;
         }
 
+        /// <summary>
+        /// Augments this builder with a custom retry ammount and delay between retries in the cases where a file lock is used.
+        /// </summary>
+        /// <param name="lockRetryDelay">Delay between retries in ms, must be 1 or more</param>
+        /// <param name="lockRetryCount">Number of retries, must be 1 or more</param>
+        /// <returns>The augmented builder</returns>
+        public StorageCreationPropertiesBuilder WithlockRetry(int lockRetryDelay, int lockRetryCount)
```
Comment (bgavrilMS) [inline src/Shared/StorageCreationPropertiesBuilder.cs]: WithLockRetry

Comment (bgavrilMS) [reply src/Shared/StorageCreationPropertiesBuilder.cs]: Nit: I'd expect there to be no retry logic unless I use this method, but it isn't the case. This method is about customizing the retry logic, not setting one up.

File: tests/Microsoft.Identity.Client.Extensions.Msal.UnitTests/MsalCacheHelperTests.cs [hunk 3]
```diff
@@ -111,21 +110,103 @@ public void MultiAccessSerializationAsync()
...
+            Stopwatch getTime = new Stopwatch();
+
+            var thread2 = new Thread(() =>
+            {
+                var args = new TokenCacheNotificationArgs(cache2, string.Empty, null, false);
+                getTime.Start();
+                try
+                {
+                    
+                    helper2.BeforeAccessNotification(args);
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Client.Extensions.Msal.UnitTests/MsalCacheHelperTests.cs]: nit:formatting

File: tests/Microsoft.Identity.Client.Extensions.Msal.UnitTests/MsalCacheHelperTests.cs [hunk 4]
```diff
@@ -111,21 +110,103 @@ public void MultiAccessSerializationAsync()
...
+                }catch(InvalidOperationException)
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Client.Extensions.Msal.UnitTests/MsalCacheHelperTests.cs]: why is an InvalidOpEx being thrown?

File: tests/Microsoft.Identity.Client.Extensions.Msal.UnitTests/MsalCacheHelperTests.cs [hunk 5]
```diff
@@ -111,21 +110,103 @@ public void MultiAccessSerializationAsync()
...
+            resetEvent2.Wait();
+
+            // Start thread 2 and give it enough time to get blocked on the lock
+            thread2.Start();
+
+            // Wait for the seconf thread to finish
+            resetEvent1.Wait();
+
+            Assert.IsTrue(getTime.ElapsedMilliseconds > 2000);
+        }
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Client.Extensions.Msal.UnitTests/MsalCacheHelperTests.cs]: Nice test

File: src/Shared/StorageCreationPropertiesBuilder.cs (hunk 2 above)
Comment (crmann1) [reply src/Shared/StorageCreationPropertiesBuilder.cs]: Updated name to be CustomizeLockRetry

File: src/Microsoft.Identity.Client.Extensions.Adal/AdalCache.cs (hunk 1 above)
Comment (crmann1) [reply src/Microsoft.Identity.Client.Extensions.Adal/AdalCache.cs]: done

**Pull request merged**